### ISSUE TITLE
HTTP Cloud Functions does not work as expected when invoked via firebase hosting rewrite rules

### ISSUE THREAD

**Issue by emadalam** [firebase/firebase-functions#858]
### Related issues

None

### [REQUIRED] Version info

**node:** v12.20.0

**firebase-functions:** 3.11.0

**firebase-tools:** 8.10.0

**firebase-admin:** 9.2.0

### [REQUIRED] Test case

Starting this morning the following setup has stopped working with the firebase hosting. Instead it started serving a default 404 page.

- Create and configure two firebase hosting site under same project (let's call them `site1` and `site2`)
- Create an HTTP Cloud Function (let's call it proxyToSite2) that simply proxies to the `site2` hosting at the default url
- Add hosting rewrite rule for path `proxy-2-site2` in `site1` firebase hosting configuration to invoke the cloud function `proxyToSite2`

Some snippets:

Cloud Function
```js
const functions = require('firebase-functions')
const httpProxy = require('http-proxy')

// the root site url for site2 hosting in firebase
const target = 'https://site2.web.app'

// the specific proxy library doesn't matter, any proxy library produces the same result
const proxy = httpProxy.createProxyServer({target, changeOrigin: true})

module.exports.proxyToSite2 = functions.https.onRequest((req, res) => proxy.web(req, res))
```

firebase.json
```json
{
  "hosting": [
    {
      "target": "site1",
      "public": "public",
      "rewrites": [
        {
          "source": "/proxy-2-site2",
          "function": "proxyToSite2"
        },
        {
          "source": "**",
          "destination": "/index.html"
        }
      ]
    },
    {
      "target": "site2",
      "public": "build",
      "rewrites": [
        {
          "source": "**",
          "destination": "/index.html"
        }
      ]
    }
  ]
}

```

### [REQUIRED] Steps to reproduce

- Do the setup as explained above
- See the response via curl or directly visiting the specific path `/proxy-2-site2` from `site1` url
- Returns the firebase default 404 page

### [REQUIRED] Expected behavior

Returns the content from `site2`

### [REQUIRED] Actual behavior

Returns the firebase hosting default 404 page.

### Were you able to successfully deploy your functions?

Yes

---
It's interesting to note that the same setup was working just fine forever, up until this morning. So it seems likely to be something related to an internal infrastructure release in Google Cloud Platform within their product orchestration that seems to have caused this issue.

**Comment by bkendall** [comment]
Hi emadalam@ - is this behavior in production, deployed to your sites, or is it in the emulators?

**Comment by PaulTR** [comment]
Hi Emad,

It looks like you also submitted this issue through Firebase Support. One of our engineers wasn't able to replicate the issue listed here with the steps given, so we'll continue this conversation in the support ticket since we'll need to grab your project ID and other information to investigate this further. Given this, I'm going to close out this GitHub issue while progress happens via Support.

Thanks!

**Comment by emadalam** [comment]
@bkendall This behaviour happens on production to the deployed sites.

@PaulTR It's weird to just close this issue though. The issue started happening since morning in two of our completely separate firebase projects (staging and production), with identical setup as described above, and no change or code deployments. Furthermore I can consistently reproduce this issue by creating a new firebase projects and following the steps mentioned.

What further information do you need to classify this as an issue? Or is this repo not the right place for these kind of issues, in which case which precise Firebase repo is fit to open this bug report. I can pretty much reproduce this issue on any new firebase project, so it doesn't seem to be related to any specific firebase project.

**Comment by kmcnellis** [comment]
@emadalam The open-source repository issues are meant for bugs in the CLI or SDKs.  We don't think this issue is related to either of those, so the official email support channel is the best place to resolve it.

**Comment by kmcnellis** [comment]
Just looping back around to this  - we did find a server-side bug, which was introduced earlier this week.  We are testing the fix now.

**Comment by bkendall** [comment]
(tis true - we try to keep service issue tracking internal)

But! The fix has been rolled out to production and you should no longer see this issue.

**Comment by emadalam** [comment]
@bkendall @kmcnellis Thanks for the responses.

The email support channels are painfully slow, and I did report to them after adding a workaround. Since it was affecting our production setup and making it unusable for most part, we had to rollout a workaround as a hotfix in the meantime. I wanted to report the issue anyway for it to get identified and acknowledged in the first place.

I haven't verified the fix, but we can't rely on the fragile nature of the Google Cloud internal orchestrations, so will give it a pass for now and maybe verify the fix to be working if in future something similar comes up. Thanks for looking into it anyways 🙌
